src/librustc_typeck/check/_match.rs

```diff
@@ -27,18 +27,96 @@ use syntax::ptr::P;
 use syntax_pos::Span;
 
 impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
-    pub fn check_pat(&self, pat: &'gcx hir::Pat, expected: Ty<'tcx>) {
-        self.check_pat_arg(pat, expected, false);
+    pub fn check_pat(&self, pat: &'gcx hir::Pat, expected: Ty<'tcx>, def_bm: ty::BindingMode) {
+        self.check_pat_arg(pat, expected, def_bm, false);
     }
 
     /// The `is_arg` argument indicates whether this pattern is the
     /// *outermost* pattern in an argument (e.g., in `fn foo(&x:
     /// &u32)`, it is true for the `&x` pattern but not `x`). This is
     /// used to tailor error reporting.
-    pub fn check_pat_arg(&self, pat: &'gcx hir::Pat, expected: Ty<'tcx>, is_arg: bool) {
+    pub fn check_pat_arg(&self, pat: &'gcx hir::Pat, mut expected: Ty<'tcx>,
+                         mut def_bm: ty::BindingMode, is_arg: bool) {
         let tcx = self.tcx;
 
-        debug!("check_pat(pat={:?},expected={:?},is_arg={})", pat, expected, is_arg);
+        debug!("check_pat(pat={:?},expected={:?},def_bm={:?},is_arg={})",
+            pat, expected, def_bm, is_arg);
+
+        let is_non_ref_pat = match pat.node {
+            PatKind::Struct(..) |
+            PatKind::TupleStruct(..) |
+            PatKind::Tuple(..) |
+            PatKind::Box(_) |
+            // PatKind::Lit(_) | // FIXME(tschottdorf): causes lots of errors
```

Comment:
OK, we gotta figure *this* out.

Reply:
I forget, did you have examples?

Reply:
Among possibly other things, we treat string literals incorrectly.

```
match s {
            "inf" => T::INFINITY,
            "NaN" => T::NAN,
            _ => { return Err(pfe_invalid()); }
        }
```

```
   Compiling core v0.0.0 (file:///Users/tschottdorf/rust/binding-modes/src/libcore)
error[E0308]: mismatched types
   --> src/libcore/num/dec2flt/mod.rs:219:13
    |
219 |             "inf" => T::INFINITY,
    |             ^^^^^ expected str, found reference
    |
    = note: expected type `str`
               found type `&'static str`

error[E0308]: mismatched types
   --> src/libcore/num/dec2flt/mod.rs:220:13
    |
220 |             "NaN" => T::NAN,
    |             ^^^^^ expected str, found reference
    |
    = note: expected type `str`
               found type `&'static str`

error[E0308]: mismatched types
   --> src/libcore/str/mod.rs:130:13
    |
130 |             "true"  => Ok(true),
    |             ^^^^^^ expected str, found reference
    |
    = note: expected type `str`
               found type `&'static str`

error[E0308]: mismatched types
   --> src/libcore/str/mod.rs:131:13
    |
131 |             "false" => Ok(false),
    |             ^^^^^^^ expected str, found reference
    |
    = note: expected type `str`
               found type `&'static str`

error: aborting due to 4 previous errors

error: Could not compile `core`.
```

Reply:
Yeah it seems like string and byte literals (and maybe slices?) are the other places this would go wrong. But I guess more generally one can have a constant with references in it (e.g., `const X: &'static u32 = &22` -- I forget if you can use such a constant in a match though.